src/pages/VeSDL/OnChainVote.tsx

```diff
@@ -0,0 +1,247 @@
+import {
+  Alert,
+  Autocomplete,
+  Box,
+  Button,
+  Link,
+  Stack,
+  TextField,
+  Typography,
+  useTheme,
+} from "@mui/material"
+import React, { useMemo, useState } from "react"
+import { commify, formatBNToString, isNumberOrEmpty } from "../../utils"
+import ArrowDownIcon from "@mui/icons-material/KeyboardArrowDown"
+import { BigNumber } from "ethers"
+import { Gauge } from "../../utils/gauges"
+import { GaugeController } from "../../../types/ethers-contracts/GaugeController"
+import VoteHistory from "./Votes"
+import { enqueuePromiseToast } from "../../components/Toastify"
+import { useActiveWeb3React } from "../../hooks"
+import { useQuery } from "@tanstack/react-query"
+import { useTranslation } from "react-i18next"
+
+interface OnChainVoteProps {
+  veSdlBalance: BigNumber
+  gauges: Partial<{
+    [lpTokenAddress: string]: Gauge
+  }>
+  gaugeControllerContract: GaugeController
+}
+export type GaugeName = {
+  [gaugeAddress: string]: {
+    address: string
+    gaugeName: string
+  }
+}
+
+export type VotesType = {
+  [gaugeAddress: string]: {
+    gaugeName: string
+    weight: BigNumber
+    voteDate: number
+  }
+}
+
+export default function OnChainVote({
+  veSdlBalance,
+  gauges,
+  gaugeControllerContract,
+}: OnChainVoteProps) {
+  const theme = useTheme()
+  const { account, chainId } = useActiveWeb3React()
+  const { t } = useTranslation()
+  const [selectedGauge, setSelectedGauge] = useState<{
+    gaugeName: string
+    address: string
+  } | null>(null)
+  const [alertMessage, setAlertMessage] = useState<string>()
+  const [voteWeightToSubmit, setVoteWeightToSubmit] = useState<string>("")
+  const [voteWeightInputError, setVoteWeightInputError] = useState<{
+    hasError: boolean
+    errorText: string
+  }>({ hasError: false, errorText: " " })
+
+  const gaugeNames = useMemo(() => {
+    const gaugeNameObj = {} as GaugeName
+    Object.keys(gauges)
+      .filter((lpTokenAddress) => !gauges[lpTokenAddress]?.isKilled)
+      .forEach((lpTokenAddress) => {
+        const gaugeName = gauges[lpTokenAddress]?.gaugeName
+        const gaugeAddress = gauges[lpTokenAddress]?.address.toLowerCase()
+        if (gaugeName && gaugeAddress)
+          gaugeNameObj[gaugeAddress] = {
+            address: gaugeAddress,
+            gaugeName: gaugeName,
+          }
+      })
+    return gaugeNameObj
+  }, [gauges])
+
+  const getVoteUsed = () => {
+    if (!account) return
+    return gaugeControllerContract["vote_user_power(address)"](account)
+  }
+  const getFilteredVotes = () => {
```

Comment:
Should these be wrapped in a `useCallback`?

Reply:
I believe `query function` don't need to be wrapped with `useCallback`.